proposals/gc/MVP.md

```diff
@@ -565,3 +750,215 @@ See [GC JS API document](MVP-JS.md) .
 * Provide functionality to generate fresh, non-canonical RTTs?
 
 * Provide a way to make data types non-eq, especially immutable ones?
+
+
+
+## Appendix: Formal Rules
+
+### Validity
+
+#### Type Indices (`C |- <typeidx> ok`)
+
+```
+C(x) = ct
+---------
+C |- x ok
+```
+
+#### Value Types (`C |- <valtype> ok`)
+
+```
+
+-----------
+C |- i32 ok
+
+C |- x ok
+-------------
+C |- ref x ok
+```
+
+...and so on.
+
+
+#### Structural Types (`C |- <strtype> ok`)
+```
+(C |- t1 ok)*
+(C |- t2 ok)*
+--------------------
+C |- func t1* t2* ok
+
+(C |- ft ok)*
+------------------
+C |- struct ft* ok
+
+C |- ft ok
+----------------
+C |- array ft ok
+```
+
+#### Defined Types (`C |- <subtype>* ok(x)`)
+
+```
+C |- st ok
+(C |- st <: expand(C(x)))*
+(x < x')*
+--------------------------
+C |- sub x* st ok(x')
+
+C |- dt ok(x)
+C |- dt'* ok(x+1)
+-------------------
+C |- dt dt'* ok(x)
+```
+
+#### Type Definitions (`C |- <deftype>* -| C'`)
+
+```
+C |- dt ok(|C|)
+---------------
+C |- dt -| C,dt
+
+x = |C|    N = |dt*|-1
+C' = C,(rec dt*).0,...,(rec dt*).N
+C' |- dt* ok(x)
+-------------------------------------
+C |- rec dt* -| C'
+
+C |- td -| C'
+C' |- td'* ok
+---------------
+C |- td td'* ok
+```
+
+#### Instructions (`C |- <instr> : [t1*] -> [t2*]`)
+
+```
+expand(C(x)) = func t1* t2*
+---------------------------------------
+C |- func.call : [t1* (ref x)] -> [t2*]
+
+expand(C(x)) = struct t1^i t t2*
+------------------------------------
+C |- struct.get i : [(ref x)] -> [t]
+```
+
+...and so on
+
+
+### Type Equivalence
+
+#### Type Indices (`C |- <typeidx> == <typeidx'>`)
+
+```
+C |- tie(x) == tie(x')
+----------------------
+C |- x == x'
+
+
+--------------------
+C |- rec.i == rec.i
+```
+
+#### Value Types (`C |- <valtype> == <valtype'>`)
+
+```
+
+---------------
+C |- i32 == i32
+
+C |- x == x'
+null? = null'?
+---------------------------------
+C |- ref null? x == ref null'? x'
+```
+
+...and so on.
+
+#### Field Types (`C |- <fldtype> == <fldtype'>`)
+
+```
+C |- t == t'
+--------------------
+C |- mut t == mut t'
+```
+
+#### Structural Types (`C |- <strtype> == <strtype'>`)
+
+```
+(C |- t1 == t1')*
+(C |- t2 == t2')*
+------------------------------------
+C |- func t1* t2* == func t1'* t2'*
+
+(C |- ft == ft')*
+----------------------------
+C |- struct ft* == struct ft'*
+
+C |- ft == ft'
+--------------------------
+C |- arrat ft == array ft'
+```
+
+#### Defined Types (`C |- <subtype> == <subtype'>`)
+
+```
+(C |- x == x')*
+C |- st == st'
+-----------------------------
+C |- sub x* st == sub x'* st'
+```
+
+### Subtyping
+
+#### Type Indices (`C |- <typeidx> <: <typeidx'>`)
+
+```
+C |- x == x'
+------------
+C |- x <: x'
+
+unroll(C(x)) = sub (x1* x'' x2*) st
+C |- x'' <: x'
```

Comment:
Would there be any downside to putting subtype depth into the syntax (optionally?) so subtype checks can be constant time rather than linear like we currently do with RTTs?

Reply:
Unless you consider subtype constraints on type imports (which is not included here), the depth is implicit in the definition, since all supertype definitions are fully transparent.

For imports, it may be worth considering some restrictions, e.g., putting the depth in the constraint, but I'd look at that separately.

Reply:
Makes sense, thanks.